### PR TITLE
Adjust loading number based on paints seen

### PR DESCRIPTION
**Pull request by jcmorrow**
This is kind of a crappy solution to the problem discussed in https://linear.app/replay/issue/DES-82/sometimes-loading-appears-stuck-at-0percent

But... also maybe it's better than what we have right now?

https://user-images.githubusercontent.com/5903784/189777158-3f11e7dc-d0fd-4853-80c6-79a56b539c7a.mp4

### CODE REVIEW

**Comment by jonbell-lot23** [comment]
Copied from Linear:

_So the worst case scenario is:

Nothing really populates on the screen

But in the corner you do see our turtle + percentage indicator, which should count up

It's possible this could be better, but I'd like to test more. Can we push live via some sort of experimental flag, or special querystring? I want to practice on a bunch of Replays to get a sense for it before it's fully live._

**Comment by jcmorrow** [comment]
OK, this branch now has an experimental flag, and so could go out whenever without affecting people.

**Comment by jcmorrow** [comment]
@bvaughn (for tomorrow) let me know how you feel about this. I kind of hate that it's not coordinated with #7732, but I'm also not sure exactly how much work it would take to get it there 🤔

File: src/ui/actions/session.ts [hunk 1]
```diff
@@ -163,6 +163,8 @@ export function createSocket(
       }
 
       const experimentalSettings: ExperimentalSettings = {
+        // Uncomment this to get a new session every time you refresh
+        // controllerKey: String(Math.floor(Math.random() * 10e8)),
```
Comment (bvaughn) [inline src/ui/actions/session.ts:167]: I wonder if it would be worth adding a URL flag that trips this? e.g. `?forceNewSession` or something?

File: src/ui/reducers/timeline.ts [hunk 2]
```diff
@@ -96,6 +94,19 @@ export const getUnprocessedRegions = (state: UIState) => state.timeline.unproces
...
+  if (state.timeline.allPaintsReceived) {
+    return 1.0;
+  }
+  const maxPaint = maxBy(state.timeline.paints, p => BigInt(p.point));
+  const maxPoint = maxBy(state.timeline.points, p => BigInt(p.point));
+  if (!maxPoint || maxPoint.point === "0") {
+    return 0.0;
+  }
+  return (1.0 * (maxPaint?.time || 0)) / maxPoint.time;
+};
```
Comment (bvaughn) [inline src/ui/reducers/timeline.ts:109]: Two thoughts:

1. This selector feels pretty heavy, given that it calls `maxBy` twice each time it's run– which loops over the entire array of paints and points. Fortunately I don't think we need to call `maxBy` since those arrays are sorted. (Can't we just look at the last element in the array?)

2. Should `getBasicProcessingProgress` be taking paints into account even if `basicProcessingLoadingBar` is disabled? (Should it not fall back to the prior behavior of comparing the latest time in the known regions?)

File: src/ui/utils/prefs.ts [hunk 3]
```diff
@@ -21,6 +21,8 @@ pref("devtools.consoleFilterDrawerExpanded", true);
 pref("devtools.hitCounts", "hide-counts");
 
 // app features
+pref("devtools.features.basicProcessingLoadingBar", false);
+pref("devtools.features.enableQueryCache", false);
```
Comment (bvaughn) [inline src/ui/utils/prefs.ts:25]: Where did "enableQueryCache" come from? 🤨 Was it missing before? It doesn't seem related to this PR.

**Review by bvaughn — APPROVED** [review]

File: src/ui/utils/prefs.ts (hunk 3 above)
Comment (jcmorrow) [reply src/ui/utils/prefs.ts:25]: Huh. It was (and still is) also on line 30 🤔  Good catch!

File: src/ui/reducers/timeline.ts (hunk 2 above)
Comment (jcmorrow) [reply src/ui/reducers/timeline.ts:109]: 1. Yes. Good point, duh! I will make that change.
2. I think either: 1) `getBasicProcessingProgress` is off, in which case this is never called, or else it is on, in which case we need paints to make any sort of reasonable estimate about how far basic processing has gotten.

File: src/ui/actions/session.ts (hunk 1 above)
Comment (jcmorrow) [reply src/ui/actions/session.ts:167]: Yeah, I have also toyed with this, but never *quite* found the right way to do it, the URL flag seems like a reasonable way to do it, though I might leave it off of this PR just because it's big enough as is, especially in combination with #7732

**Comment by bvaughn** [comment]
Seems like this was merged with a TSC error?
```
src/ui/setup/dynamic/devtools.ts(314,5): error TS2322: Type '(...args: any[]) => FocusRegion | null | undefined' is not assignable to type 'SelectorWithoutStateArg<ActionCreatorWithoutPayload<string>> & SelectorWithoutStateArg<(state: UIState) => boolean> & ... 204 more ... & SelectorWithoutStateArg<...>'.
  Type '(...args: any[]) => FocusRegion | null | undefined' is not assignable to type 'SelectorWithoutStateArg<ActionCreatorWithoutPayload<string>>'.
    Type 'FocusRegion | null | undefined' is not assignable to type '{ payload: undefined; type: string; }'.
      Type 'undefined' is not assignable to type '{ payload: undefined; type: string; }'.
src/ui/setup/dynamic/devtools.ts(31[5](https://github.com/replayio/devtools/actions/runs/3065125050/jobs/4948906217#step:6:6),42): error TS255[6](https://github.com/replayio/devtools/actions/runs/3065125050/jobs/4948906217#step:6:7): A spread argument must either have a tuple type or be passed to a rest parameter.
Error: Process completed with exit code 2.
```

File: src/ui/utils/timeline.ts [hunk 4]
```diff
@@ -309,3 +310,48 @@ export function filterToFocusRegion<T extends TimeStampedPoint>(
...
+
+function assertSorted(a: TimeStampedPoint[]) {
+  a.reduce(
+    (prev, curr) => {
+      assert(prev.time <= curr.time);
+      return curr;
+    },
+    { point: "0", time: 0 }
+  );
+}
```
Comment (bvaughn) [inline src/ui/utils/timeline.ts:322]: I'm seeing this assertion throw errors for recording 436b7cf5-50f1-4e69-894d-a61343e128c2 but it's not really clear how to use that information.

Comment (bvaughn) [reply src/ui/utils/timeline.ts:322]: I _think_ this assert actually uncovered a bad recording (see BAC-2339).

I'm going to change it for now though to (1) only run in DEV and (2) compare points rather than time (since those are more fine-grained anyway). Let's chat if you think this is a bad move!

File: src/ui/utils/timeline.test.ts [hunk 5]
```diff
@@ -231,3 +232,32 @@ describe("filterToFocusRegion", () => {
     ]);
   });
 });
+
+describe("mergeSortedPointLists", () => {
+  fit("will correctly interleave overlapping ranges", () => {
```
Comment (bvaughn) [inline src/ui/utils/timeline.test.ts:237]: We should probs add a line rule to prevent `fit` or `test.only` from being committed accidentally. I do this sometimes too.
